compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGenOptimizationReportEmitter.java

```diff
@@ -0,0 +1,172 @@
+ *
+ *  WSO2 LLC. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.projects;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import org.ballerinalang.model.elements.PackageID;
+import org.ballerinalang.model.symbols.SymbolOrigin;
+import org.wso2.ballerinalang.compiler.bir.model.BIRNode;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.PrintStream;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.wso2.ballerinalang.compiler.util.ProjectDirConstants.TARGET_DIR_NAME;
+
+/**
+ * Emits time durations and optimized node details for codegen optimization.
+ *
+ * @since 2201.10.0
+ */
+public class CodeGenOptimizationReportEmitter {
+
+    private static final Map<PackageID, Long> birOptimizationDurations = new LinkedHashMap<>();
+    private static long nativeOptimizationDuration = 0;
+    private static final PrintStream out = System.out;
+    private static final String CODEGEN_OPTIMIZATION_REPORT = "codegen_optimization_report.json";
+
+    protected static void flipBirOptimizationTimer(PackageID pkgId) {
+        if (birOptimizationDurations.containsKey(pkgId)) {
+            long totalDuration = System.currentTimeMillis() - birOptimizationDurations.get(pkgId);
+            birOptimizationDurations.put(pkgId, totalDuration);
+        } else {
+            birOptimizationDurations.put(pkgId, System.currentTimeMillis());
+        }
+    }
+
+    protected static void flipNativeOptimizationTimer() {
+        if (nativeOptimizationDuration == 0) {
+            nativeOptimizationDuration = System.currentTimeMillis();
+        } else {
+            nativeOptimizationDuration = System.currentTimeMillis() - nativeOptimizationDuration;
+        }
+    }
+
+    protected static void emitBirOptimizationDuration() {
+        long totalDuration = birOptimizationDurations.values().stream().mapToLong(Long::longValue).sum();
+        out.printf("Duration for unused BIR node analysis : %dms%n", totalDuration);
+        birOptimizationDurations.forEach((key, value) -> out.printf("    %s : %dms%n", key, value));
+        out.println();
+        // TODO make this class a singleton
```

Comment:
Are you having this TODO as a future improvement? If not, let's remove this.

Reply:
The cleaner way is making the class a singleton instead of cleaning static variables after each run. I will change it and update here.

Reply:
Fixed in 39f77f3adabd723a8733addcd08e954a0db6cd3e.